flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/GlobalStatisticsTracker.java

```diff
@@ -0,0 +1,120 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.shuffle;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * GlobalStatisticsTracker is used by {@link DataStatisticsCoordinator} to track the in progress
+ * {@link GlobalStatistics} received from {@link DataStatisticsOperator} subtasks for specific
+ * checkpoint.
+ */
+@Internal
+class GlobalStatisticsTracker<D extends DataStatistics<D, S>, S> {
+  private static final Logger LOG = LoggerFactory.getLogger(GlobalStatisticsTracker.class);
+  private static final double EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE = 90;
+  private final String operatorName;
+  private final TypeSerializer<DataStatistics<D, S>> statisticsSerializer;
+  private final int parallelism;
+  private volatile GlobalStatistics<D, S> inProgressStatistics;
+
+  GlobalStatisticsTracker(
+      String operatorName,
+      TypeSerializer<DataStatistics<D, S>> statisticsSerializer,
+      int parallelism) {
+    this.operatorName = operatorName;
+    this.statisticsSerializer = statisticsSerializer;
+    this.parallelism = parallelism;
+  }
+
+  GlobalStatistics<D, S> receiveDataStatisticEventAndCheckCompletion(
+      int subtask, DataStatisticsEvent<D, S> event) {
+    long checkpointId = event.checkpointId();
+
+    if (inProgressStatistics != null && inProgressStatistics.checkpointId() > checkpointId) {
+      LOG.debug(
+          "Expect data statistics for operator {} checkpoint {}, but receive event from older checkpoint {}. Ignore it.",
+          operatorName,
+          inProgressStatistics.checkpointId(),
+          checkpointId);
+      return null;
+    }
+
+    GlobalStatistics<D, S> completedStatistics = null;
+    if (inProgressStatistics != null && inProgressStatistics.checkpointId() < checkpointId) {
+      if ((double) inProgressStatistics.aggregatedSubtasksCount() / parallelism * 100
+          >= EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE) {
+        completedStatistics = inProgressStatistics;
+        LOG.info(
+            "Received data statistics from {} operators {} out of total {} for checkpoint {}. "
+                + "It's more than the expected percentage {}. Complete data statistics aggregation {}",
+            inProgressStatistics.aggregatedSubtasksCount(),
+            operatorName,
+            parallelism,
+            inProgressStatistics.checkpointId(),
+            EXPECTED_DATA_STATISTICS_RECEIVED_PERCENTAGE,
+            completedStatistics);
+      } else {
+        LOG.info(
+            "Received data statistics from {} operators {} out of total {} for checkpoint {}. "
+                + "It's less than the expected percentage {}. Dropping the incomplete aggregate "
```

Comment:
similar to above `Aborting the incomplete aggregation for checkpoint {} and starting a new aggregation for checkpoint {}`